src/Container/ApplicationConfigInjectionDelegator.php

```diff
@@ -0,0 +1,373 @@
+        // in the order in which they are inserted.
+        $serial = PHP_INT_MAX;
+        return function ($queue, $item) use (&$serial) {
+            $priority = isset($item['priority']) && is_int($item['priority'])
+                ? $item['priority']
+                : 1;
+            $queue->insert($item, [$priority, $serial]);
+            $serial -= 1;
+            return $queue;
+        };
+    }
+
+    /**
+     * Pipe routing middleware into the application.
+     *
+     * Attempts to pull both the router and response interface from the application
+     * container, falling back to reflection of the application instance
+     * when unable to find either one.
+     *
+     * @return void
+     */
+    private static function pipeRoutingMiddleware(Application $application)
+    {
+        $container = $application->getContainer();
+        if ($container->has(RouteMiddleware::class)) {
+            $application->pipe(RouteMiddleware::class);
+            return;
+        }
+
+        $application->pipe(new RouteMiddleware(
+            self::getRouter($application, $container),
+            self::getResponsePrototype($application, $container)
+        ));
+    }
+
+    /**
+     * Pipe dispatch middleware into the application.
+     *
+     * If the DispatchMiddleware is present in the application's container, it
+     * pipes that service. Otherwise, instantiates it directly and pipes it.
+     *
+     * @return void
+     * @throws MissingDependencyException if the RouterInterface service is not
+     *     found when the RouteMiddleware is not present.
+     */
+    private static function pipeDispatchMiddleware(Application $application)
+    {
+        $container = $application->getContainer();
+        $application->pipe(
+            $container->has(DispatchMiddleware::class)
+            ? $container->get(DispatchMiddleware::class)
+            : new DispatchMiddleware()
+        );
+    }
+
+    /**
+     * Get the router.
+     *
+     * If not available in the container, uses reflection to pull it from the
+     * application.
+     *
+     * @return RouterInterface
+     */
+    private static function getRouter(Application $application, ContainerInterface $container)
+    {
+        if (! $container->has(RouterInterface::class)) {
+            $r = new ReflectionProperty($application, 'router');
+            $r->setAccessible(true);
+            return $r->getValue($application);
+        }
+
+        return $container->get(RouterInterface::class);
+    }
+
+    /**
+     * Get the response prototype.
+     *
+     * If not available in the container, uses reflection to pull it from the
+     * application.
+     *
+     * If in the container, fetches it. If the value is callable, uses it as
+     * a factory to generate and return the response.
+     *
+     * @return ResponseInterface
+     */
+    private static function getResponsePrototype(Application $application, ContainerInterface $container)
```

Comment:
This method is only necessary in v2, as the response prototype may be injected only in the application, and not the container.